dev-environment.yml

```diff
@@ -21,14 +21,15 @@ channels:
 dependencies:
   # Project dependencies
   - confuse
+  - jpype1 >= 1.4.1
```

Comment:
It would be unwise to use primitive and array type directly if you are trying to support two backends unless those classes are guaranteed to be fully compatible.  I know very little about jep implementation, but they aren't likely to have nearly the same buffer/memoryview interfacing.  

Technically JPype handles can be handed to  different implementations. And as far as I know it supports monkey patching given that JPype even gives you a monkey patch API for Java classes.    It is just the wrapper for the jvalue that defines the type.  So if scyjava wants to define its own base classes it is not prohibited.  However if you want to do that you woild likely need to push for JPype to add the proposed return hooks which allow one user control on the packaging of return values.  Noone ever seconded that feature so like the reverse bridge it is waiting with no user need.

If scyjava wants to define base types on subclassing it would need to pass the constructor key.  To prevent the user from accidently trying to extend a Java type a special key must be passed.  All JPype objects are split into a C backend and a Python front end for speed.  You can customize the front end as much as  you desire.  Monkey patching backends may be harder without mro tricks as the CPython doesnt allow much modification of C types.  But we didnt close the class and they are heep types so anything that Python allows is doable.

Much of these details are in the developers guide.

Reply:
> Unfortunately still no volunteers for the 120 hours of writing test benches that would finish epypj so that jpype can do both forward and reverse.

@Thrameos Could you point me at some resources for understanding what you mean? Mailing list or forum threads, or elaborate a little on what this would entail? My team could potentially help with this depending on how much work is required, since we have a definite need for the Python-inside-Java direction, and would like to help JPype achieve this, but we are not expert on low-level JNI and Cython technologies.

Reply:
@Thrameos I took an initial look at the epypj branch. Your aspirations are greater than what we have been trying to achieve with scyjava+jep so far, in that we aren't trying to wrap lots of Python types to Java objects. We mainly just need to run the same Python scripts that pull from the Java side, regardless of whether it's Java-inside-Python or Python-inside-Java. So for testing, what I did for scyjava+jep was to call exactly the same Python test suite in both cases (minus a couple of non-working rough edges e.g. Java class reflection in Python via Jep). We were able to get all our tests passing in both cases, where the Python-inside-Java cases uses Jep. That said, I think the Java-side wrappers for Python types are great, but if JPype simply worked the same way on the Python side, that would already be a big win. Do you think it would make sense to adapt/reuse (much of) the `test/jpypetest` suite for this case?

If the Java-side Python type wrappers were to be completed, would you then also want them to work in the Java-inside-Python case? And would you want the test suite you started working on in `test/testng/epypj` to also work for Java-inside-Python?

Reply:
> For some of these things, it might be enough to have scyjava accessors returning the classes with correct behavior. So e.g. `scyjava.JArray` would point to `jpype.JArray` or `jep.PyJArray` depending on mode.

Well, I'd say this problem is beyond the scope of this PR. If you want to file an issue for it (or in pyimagej) I'd be happy to work on it at some point.

Reply:
@ctrueden I think that we don't have to go to the full epypj for what you are looking for (at least based on your description).   It is easy to set up a system in regular JPype in which you start the JVM and launch Python from the CAPI.  You then define a class as Java interface with all the calls you want to be able to make to Python (run script, exec, get/set global variables).  A Python script gets run, calls JImplements which installs all the hooks needed, passes the Python instance handle back to Java and tada you have a fully functional interface to Python initiated from the Java side.   The early prototype of epypj I wrote at first easily covered all that type of functionality.  Don't know if I still have that branch around, but it was only a few hours of work to construct.

Epypj is definitely much larger which maybe why I couldn't ever find any users.   It was intended to make all Python classes automatically usable within Java under the ``python.lang`` using on the fly generated wrappers.  Secondarily it was to have a precompiler which would take a Python package and compile the forward declarations for it so that Java could use it natively.   Thus you could then have Java code that would be able to use numpy and matplotlib no differently that any other Java package other that the inconvenience of having to deal with the weak typing of Python while working with Java.   It does change the nature of JPype a bit because now passing a Python list to a Java list would not convert but instead give up the ``python.lang.PyList`` handle containing a list of ``python.lang.PyObject`` instances, etc.  Meaning there would be no assumption of conversion except for concrete types.

My sticking point was after I wrote all the dynamic code, I needed a new test framework like testng and coverage tool so that we could get all of those new entry points tested.   "Can you call a PyList returning method, can you access the nth element, what happens if you access an index that is too far in the list, what it there is an exception, etc"   It took many hours and thousands of tests to just get the one side of the bridge working, but unless I could find a group of users that could contribute to that testing it would be very hard to justify it myself as my paid job uses the bridge only Py->J.  The other sticking point is that if one uses the bridge both ways, it is very easy to create irresolvable memory loops that neither Python nor Java can resolve.   Currently as a one way bridge it is very hard to stuff a Python object into a Java list which points back to the Python object.   (It is still possible if you use JProxy to pass a Python object as a ``Serializable`` instance.)  Once I make it a two way bridge it is easy.  Every Java container that you can place a Python object into or Python container that you stuff Java object into is a memory leak.  And as Python objects have lots of hidden connections, it may be hard to even know you are doing it.   I even started a new virtual machine which was able to handle foreign objects as proof that such a system could exist (which could in principle have run Java and Python), but it was too large a project to complete by myself.     But I digress...

The remaining task for epypj is to create a Java centric testbench which much like the jpypetest would hit every aspect of each of the exposed classes such that we had in excess of 80% coverage (and 100% of the FFI calls) at which point it could be released as jpype 2.0.  We don't have to solve the memory issue as we are warning the users that they should use the bridge in one direction at a time (Python to Java  or Java to Python but not mixed).